libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/MapboxAudioGuidance.kt

```diff
@@ -126,8 +128,8 @@ internal constructor(
     private fun audioGuidanceFlow(
         mapboxNavigation: MapboxNavigation
     ): Flow<MapboxAudioGuidanceState> {
-        return mapboxNavigation.audioGuidanceVoice().flatMapLatest { audioGuidance ->
-            var lastPlayedInstructions: VoiceInstructions? = null
+        var lastPlayedInstructions: VoiceInstructions? = null
+        return mapboxNavigation.audioGuidanceVoice().flatMapMerge { audioGuidance ->
```

Comment:
Why do you use `flatMapMerge` now? With it we may handle multiple `MapboxAudioGuidanceVoice` instances in parallel, that doesn't sound correct to me.

Reply:
That's a weird thing here. If we cancel everything once we get a new `MapboxAudioGuidanceVoice` instance, we get the issue with playing the same instruction twice or starting to play the first one and then interrupting it and playing the second one.
Capturing from https://mapbox.atlassian.net/browse/NAVAND-1001?focusedCommentId=76093:

> 1) We are in FreeDrive with language=en and `MapboxAudioGuidanceVoice#1`.
> 2) We receive new routes via `RoutesObserver`.
> 3) We receive first voice instruction via `VoiceInstructionsObserver`.
> 4) We play it using `MapboxAudioGuidanceVoice#1`.
> 5) The voice changes due to routes update in p. 2 (the collect callback is invoked [here](https://github.com/mapbox/mapbox-navigation-android/blob/8dab187b06cf0786553777655658afcb070ec8f6/libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/MapboxAudioGuidance.kt#L171)).
> 6) We have language=en-US and `MapboxAudioGuidanceVoice#2`.
> 7) `MapboxAudioGuidanceVoice#2` gets current voiceInstructionsFlow value and plays it.
> 
> My guess is that the order is wrong because in one place we use combine (when creating language based on routes) and in the other we observe the flow directly (for voice instructions).
> Confirmed: the issue is gone when not using combine.
> We should look for a good way to keep the order guaranteed. Either that or we should keep `lastPlayedInstructions` across `MapboxAudioGuidanceVoice` instances. However, that raises the question: if the language changes, do we want to repeat the latests instruction in the other language?

And with this solution we won't run into this situation: we'll first finish playing the first instruction with `MapboxAudioGuidanceVoice#1`, then try to play the second one with `MapboxAudioGuidanceVoice#2` but it will get filtered out because it's saved in `lastPlayedInstructions`.

There are other possible solutions. For example, I don't like it that language changes between FreeDrive and ActiveGuidance. That's the reason `MapboxAudioGuidanceVoice` instance changes. I've tried to bring it up but people considered it expected.
Another way to go would be to guarantee that the first voice instruction will be delivered strictly after the corresponding language change. But I wasn't able to find a way to do it with the current paradigm.

Reply:
I think the wrong order can be fixed by using `Dispatchers.Main.immediate` instead of `Dispatchers.Main` when creating a `MapboxAudioGuidance`. At least my logs tell me so.